Detsim city flow

gondiaz opened the pull request:
This is the last of the series of PRs #757 #756 #695  #694.

It contains detsim city flow module that performs a fast detector simulation.

From MC hits, it simulates inonization electron creation, drift and diffusion. Then it simulates the photon propagation based on previously computed light-tables, which depend on detector geometry. The final output are the signal waveforms at sensors (without electronic simulation).

mmkekic commented on invisible_cities/cities/detsim.py:
```diff
@@ -0,0 +1,216 @@
+
+# @profile
+@city
+def detsim(files_in, file_out, event_range, detector_db, run_number, s1_lighttable, s2_lighttable, sipm_psf,
```
you should probably group input parameters in dictionaries

gondiaz replied:
I grouped in `physics_parameters` and `buffer_parameters`. Maybe we can also group ligth tables paths.

mmkekic commented on invisible_cities/cities/detsim.py:
```diff
@@ -0,0 +1,216 @@
+    return select_hits
+
+
+def ielectron_simulator(wi: float, fano_factor: float, lifetime: float,
```
Maybe make them keyword-only arguments to avoid possible misuse?

gondiaz replied:
Maybe this is solved by the` physics_parameters` grouping.

mmkekic commented on invisible_cities/cities/detsim.py:
```diff
@@ -0,0 +1,216 @@
+            function that returns the positions emission times and number of photons at the EL
+    """
+    def simulate_ielectrons(x, y, z, time, energy):
+        electrons = generate_ionization_electrons(energy, wi, fano_factor)
```
How about `num_electrons`, or `nelectrons` so it is clear you are getting the number of electrons?

mmkekic commented on invisible_cities/cities/detsim.py:
```diff
@@ -0,0 +1,216 @@
+    return get_buffer_times_and_length
+
+
+def s2_waveform_creator (sns_bin_width, LT, EL_drift_velocity):
```
use lowercase for variables (uppercase for classes instances are ok)

jacg replied:
If we are used to seeing it upper case in our standard notation outside of source code, I would recommend sticking to uppercase even in source code.

But we do follow PEP-8 on *no-spaces-before-parentheses*, unless there's an overriding reason (usually vertical alignment) to do otherwise: I don't see such a reason here.

mmkekic commented on invisible_cities/cities/detsim.py:
```diff
@@ -0,0 +1,216 @@
+    return get_bin_edges
+
+
+# @profile
```
remove

mmkekic reviewed (commented):
First round of mostly style comments

jacg commented on invisible_cities/cities/detsim.py:
```diff
@@ -98,39 +98,29 @@ def create_s2_waveform(xs, ys, ts, phs, tmin, buffer_length):
            detector_db, run_number, s1_lighttable, s2_lighttable, sipm_psf,
-           buffer_parameters  = dict(),
-           physics_parameters = dict()):
+           buffer_params  = dict(),
```
Eeeeek! Don't use mutable objects as default arguments!

(Do you know how to deal with this, or shall I explain?)

gondiaz replied:
Please explain. Is not the same case in [dorothea](https://github.com/next-exp/IC/blob/master/invisible_cities/cities/dorothea.py), [penthesilea](https://github.com/next-exp/IC/blob/master/invisible_cities/cities/penthesilea.py) and [esmeralda](https://github.com/next-exp/IC/blob/master/invisible_cities/cities/esmeralda.py)?

jacg replied:
The fundamental problem is that (in Python) default arguments are evaluated only once, at the time the function is defined. Consequently, the same *single object* is shared by all calls to the function which do not provide an overriding argument. If this object is mutated in any of those calls, then all subsequent calls use a different default value from the one visible in the function's parameter list. When this happens, the resulting observations (if you spot them; you might not, and then you're producing incorrect results) tend to be so freakishly weird, and waste so much debugging time, that we simply want to make sure that we never get to that stage. So, we studiously avoid *ever* using mutables as default arguments, directly.

Try this simple experiment to see it in action:

``` python
def foo(x, a=[]):
    a.append(x)
    return max(x, *a)

foo(1)   # 1
foo(666) 
foo(1)   # 666
```

It's a subtle manifestation of the scourge of mutable shared state.

A simple solution would be to use a frozen dict. Unfortunately a PEP proposing frozen dicts was rejected some years ago, leaving us with the bizarre situation that `frozenset` is a builtin, while its dictionary brother was not allowed in.

This leaves two obvious options:

1. Use the frozen dict package. Web search left as an exercise. I've no experience with it, so I can't vouch for it.

2. Use the standard idiom for dealing with mutable default arguments:

   ```python
   def fn(parameter=None):
       if parameter is None:
           parameter = <create *NEW INSTANCE* of mutable default value here>
   ```

   which is a bit annoying, but always works.
   
> Is not the same case in dorothea, penthesilea and esmeralda?

If so, then it should be fixed there too!

Not necessarily in this PR. If there is someone new to Next/IC/Git/PRs/our-workflow, nervously waiting in the wings, then fixing this problem in those other places could be an excellent introduction to submitting a PR in our environment.

jacg replied:
Oh ...

> Is not the same case in dorothea, penthesilea and esmeralda?

... let this act as an illustration of why the practice of copy pasting stuff without understanding what it means and why it is done the way it is, has significant costs and is thoroughly discouraged.

Invest time and effort in understanding things deeply, and it will pay dividends in the long term.

gonzaponte replied:
But this is a city, i.e. this function will never be called twice in the same program. At least it is not meant to. Isn't that a good motivation to reduce the noise introduced by initializing them to `None`? On the other hand, I don't know why these are the only parameters that are initialized to "nothing" as if they were optional, aren't they essential to the running of the city?

gondiaz replied:
Yes, they are essential. Those parameters being an empty dictionary as they are by default won't work. The config file must contains all the parameters defined.

jacg replied:
> this function will never be called twice in the same program. 

You hope ...

> At least it is not meant to.

... there you go: things than are not meant to happen, tend to have a habit of happening anyway, every now an then at least.

> Isn't that a good motivation to reduce the noise introduced by initializing them to None?

It may be.

A competent Python programmer should get very nervous at seeing mutable default values, and will then have to waste time wondering about the safety, or fixing the (non-)issue. So, if it stays, it must be accompanied by comments as to why it is safe (which the reader may, or may not, trust) ... at which point it might be less noisy simply to do the `None` trick.

jacg replied:
> Those parameters being an empty dictionary as they are by default won't work. 

So the default value defined in the function must *always* be overridden for the function to stand a chance of working?

That's so wrong!

Why not turn them into keyword-only args?

jacg replied:
> Why not turn them into keyword-only args?

... in fact, IIRC, it would simply make sense to make all parameters of all cities keyword-only.

But, even without that, just remove the default values, and everything should work, shouldn't it?

gonzaponte replied:
This is something inherent to all cities in IC. All arguments are actually handled as keyword arguments by the `city` decorator (IIRC). We could turn them into keyword-only arguments, and maybe we should as I don't think it is of any use to have positional arguments. Somehow it bothers me to have that `*` at the beginning of the function signature, but it is definitely something irrelevant for the decision.

jacg replied:
My hypotheses are

1. Adding `*` to the front of each city's signature should break nothing.
2. Removing all `= dict()` from all cities' signatures should break nothing.

I'm pretty confident on the first: that's by design of the whole liquid city thing. If anything breaks, it will be because, someone, somewhere is abusing something. Adding the `*` reduces the chances of someone abusing something in the future.

I'm less confident on the second: these are lower-level details so I'd have to look into the details case-by-case.

I propose removing the `= dict()`s in this city, and running all the tests. If that works, the obvious followups are the two points listed above, in reverse order.

gonzaponte replied:
I agree. The first point will work for sure. The second one might break some random tests that don't care about the dict-like arguments to run, but it won't break any realistic usage of the city.

mmkekic commented on invisible_cities/cities/detsim_test.py:
```diff
@@ -0,0 +1,103 @@
+
+    # modify labels for event_id = 0 in test file
+    # all its hits labels are set to "NO_ACTIVE"
+    modified_testfile = os.path.join(config_tmpdir, "modified_testfile.h5")
```
the files are less than 100KB, is probably ok to include them in the standard test suite

mmkekic commented on invisible_cities/cities/detsim_test.py:
```diff
@@ -0,0 +1,103 @@
+    PATH_OUT = os.path.join(config_tmpdir, "detsim_test.h5")
+    conf = configure('detsim $ICTDIR/invisible_cities/config/detsim.conf'.split())
+
+    buffer_length     = 800 * units.mus
```
You seem to be reusing `conf['buffer_length']` rather than changing all its values, so you can either update it directly or create a local reference to it and update it since those are all mutable objects

jacg replied:
> create a local reference to it and update it since those are all mutable objects

Didactic point ... I think that 'local reference' is an unfortunate choice of words: you should have said 'local copy'. 'Local reference' is not strictly incorrect, but it's ambiguous and fails to convey the important point. I'd go as far as saying that its most obvious interpretation is the exact opposite of what you mean!

It would be good if you all developed a habit of getting nervous when something is mutated, especially something that you got from somewhere else. Mutable shared state is the root of all evil in programming. Sometimes we need to mutate for pragmatic reasons, but do avoid mutation wherever pragmatically possible. In this case, we're talking about copying a smallish dictionary in a test that does IO (hint: IO is veeeeeeery slow compared to in-memory operations).

If you make a local copy of this dict, you can mutate it to your heart's content, and nobody ever has to worry that you're breaking things in some other part of the code which holds another reference to the thing you are mutating.

mmkekic commented on invisible_cities/cities/detsim_test.py:
```diff
@@ -0,0 +1,103 @@
+from invisible_cities.core  import system_of_units as units
+
+
+def test_detsim_contains_all_tables(ICDATADIR, config_tmpdir):
```
probably you want to use `output_tmpdir` (yes, i know in other city tests is usually all placed in config one); also is better to name the output files differently so you are not overwriting the same file in one pytest run

mmkekic commented on invisible_cities/cities/detsim_test.py:
```diff
@@ -0,0 +1,277 @@
+    PATH_OUT = os.path.join(output_tmpdir, "detsim_test.h5")
+    conf = configure('detsim $ICTDIR/invisible_cities/config/detsim.conf'.split())
+
+    buffer_params = dict()
```
why not 
`dict(length=800*mus, pmt_width=100*units.ns`...?

mmkekic commented on invisible_cities/cities/detsim_test.py:
```diff
@@ -0,0 +1,277 @@
+    # modify labels for event_id = 0 in test file
+    # all its hits labels are set to "NO_ACTIVE"
+    modified_testfile = os.path.join(output_tmpdir, "modified_testfile.h5")
+    shutil.copyfile(PATH_IN, modified_testfile)
```
Just add the modified file to the test files

mmkekic commented on invisible_cities/cities/detsim_test.py:
```diff
@@ -0,0 +1,277 @@
+    PATH_OUT = os.path.join(output_tmpdir, "detsim_test.h5")
+    conf = configure('detsim $ICTDIR/invisible_cities/config/detsim.conf'.split())
+
+    # remove test file rows
```
again, add empty file to the test files

mmkekic commented on invisible_cities/cities/detsim_test.py:
```diff
@@ -0,0 +1,277 @@
+    PATH_TRUE = os.path.join(ICDATADIR    , "detsim_new_kr83m_fast.newformat.sim.h5")
+    conf = configure('detsim $ICTDIR/invisible_cities/config/detsim.conf'.split())
+
+    buffer_params = dict()
```
No need to initialize empty dictionary, just create the dictionary with all the values you need.  You could **copy** the  `conf['buffer_params']`already existing dictionary and modify it since the only modification seems to be `pmt_width`

mmkekic commented on invisible_cities/cities/detsim_test.py:
```diff
@@ -0,0 +1,277 @@
+    PATH_IN   = os.path.join(ICDATADIR    , "nexus_new_kr83m_fast.newformat.sim.h5")
+    PATH_OUT  = os.path.join(output_tmpdir, "detsim_test.h5")
+
+    physics_params = dict()
```
Again, copy already existing dictionary inside `conf` and only modify needed values

mmkekic commented on invisible_cities/cities/detsim_test.py:
```diff
@@ -0,0 +1,277 @@
+    buffer_params["trigger_thr"] = 0
+
+    # copy to avoid mutability
+    physics_params_ = physics_params.copy()
```
I dont see that you later plan to modify `physics_params` so why copying it again?

gondiaz replied:
`physics_params_` is mutted inside detsim city (`ws` and `el_drift_velocity`), therefore the second city call can not be done with the same dictionary. They will take the values from **detsim.conf** from now on.

mmkekic commented on invisible_cities/cities/detsim_test.py:
```diff
@@ -0,0 +1,277 @@
+
+    # modify test file
+    modified_testfile = os.path.join(output_tmpdir, "modified_testfile.h5")
+    shutil.copyfile(PATH_IN, modified_testfile)
```
Add the `modified_testfile.h5`to test files (and rename it)

mmkekic commented on invisible_cities/cities/detsim_test.py:
```diff
@@ -0,0 +1,277 @@
+                assert_tables_equality(got, expected)
+
+
+def test_detsim_buffer_times(ICDATADIR, output_tmpdir):
```
Add a comment on what this test does, for example the nonphysical `el_drift_velocity`is confusing. Also, again,  you can copy dictionaries from conf and update the changed parameters

mmkekic commented on invisible_cities/cities/detsim_test.py:
```diff
@@ -0,0 +1,277 @@
+    conf.update(dict(files_in      = PATH_IN,
+                     file_out      = PATH_OUT,
+                     run_number    = 0,
+                     event_range   = (1, 2),
```
Can you add a comment why you skip the first event from the file? I understand it is because Z is very small for the first event in the file

mmkekic commented on invisible_cities/cities/detsim_test.py:
```diff
@@ -0,0 +1,277 @@
+    s2_tmax = max(s2_times) + buffer_params["pre_trigger"]
+
+    with tb.open_file(PATH_OUT) as h5out:
+        pmtwf  = h5out.root.pmtrd.read()[0].sum(axis=0)
```
Can you also read sipmwfs (so we check timings there too)

mmkekic commented on invisible_cities/cities/detsim_test.py:
```diff
@@ -0,0 +1,277 @@
+    # start s1 time
+    tstart = signal_bins[0]*buffer_params["pmt_width"]
+    assert tstart == buffer_params["pre_trigger"]
+
```
Maybe now you can simply assume that s2 will be after `pre_trigger + 2*mus` so set to s2s indices whatever is after that time. Then you can check that both start and end make sense. And do the same for sipms waveforms (without the s1 of course)

mmkekic commented on invisible_cities/cities/detsim.py:
```diff
@@ -0,0 +1,214 @@
+    lt_sipm = LT_SiPM(fname=os.path.expandvars(sipm_psf), sipm_database=datasipm)
+    el_gap = lt_sipm.el_gap_width
+
+    select_s1_candidate_hits = fl.map(hits_selector(False),
```
This is not even used in the flow!

mmkekic commented on invisible_cities/cities/detsim.py:
```diff
@@ -0,0 +1,214 @@
+from .. detsim.light_tables_c     import LT_SiPM
+from .. detsim.light_tables_c     import LT_PMT
+from .. detsim.s2_waveforms_c     import create_wfs
+from .. detsim.detsim_waveforms   import s1_waveforms_creator
```
Remove all unnecessary imports (use pyflakes to discover which ones to remove)

mmkekic reviewed (commented):
After some commits squashing I am ready to approve this PR

gondiaz committed:
Add detsim city flow module and config file

gondiaz committed:
Add detsim city test module and test files

The tests include:
- contains all tables
- filter
- empty input
- exact result
- exact time translation
- correct buffer times

mmkekic reviewed (approved):
This PR is the last piece of the long lasting effort to implement fast detector light simulation in IC. The PR introduces the final city flow that is using already approved functionalities. Code is readable and appropriate tests are added. Good job!
